### ISSUE TITLE
Disable invalid_annotation_target if used on a class annotated by @freezed

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#2778]
The code-generator [Freezed](https://github.com/rrousselGit/freezed) works by transposing annotations, such that we have:

```dart
@freezed
@a
class Example with _$Example {
  @b
  factory Example(@c int value) = _Example;
}
```

somehow generates:

```dart
@a
class _Example implements Example {
  @b
  _Example(this.value);

  @c
  final int value;
}
```

The problem is, this lead to invalid_annotation_target warnings if users use an annotation annotation with `@Target`.

Which means that the recent versions of meta/analyzer could require users to disable the warning

Would it be feasible to somehow detect that the class is annotated with `@Freezed` from `package:freezed_annotation/freezed_annotation.dart` and disable the warning?

**Comment by bwilkerson** [comment]
Before I attempt to answer that question I want to make sure that I understand the problem, so please let me know if I'm wrong at some point. I'm guessing that
- there aren't any diagnostics in the original code and that the unwanted diagnostic is in the generated code,
- `@a` doesn't generate any diagnostic in the generated code because in both cases it's on a class, and
- similarly `b` doesn't generate any diagnostic in the generated code because in both cases it's on a constructor.

Therefore I'm guessing that the problem is that the class of `@c` is annotated with `@Target` to be valid on a parameter but not on a field.

I'm also guessing that you don't control either `Freezed` nor the definition of `@c` because if you did you could either fix the generator to be smarter about copying over annotations or fix the definition of `@c`'s type to allow it on fields. And I assume that `Freezed` doesn't copy across comments, so you can't just add an ignore comment in the original code that would then prevent the diagnostic from being reported in the generated code.

> Would it be feasible to somehow detect that the class is annotated with `@Freezed` ...

Unless you forgot to copy the `@freezed` annotation when you pasted the generated code above it isn't there for the analyzer to detect.

If it is there, then I'd be very hesitant to add special knowledge of packages that aren't controlled by either the Dart or Flutter teams to the tooling because it's too easy for that knowledge to get out of sync with reality.

Is there a reason the generated code isn't being excluded from analysis? I don't work with generated code, but it's my understanding that most generated code isn't analyzed.

**Comment by rrousselGit** [comment]
> there aren't any diagnostics in the original code and that the unwanted diagnostic is in the generated code,

There is. The diagnostics are in the user defined code: The `Example` class, not `_Example` (although my snippet has a mistake, which I fixed)

You are correct that `@a` does not cause problems. On the other hand, b and c do.


> I'm also guessing that you don't control either Freezed nor the definition of @c because if you did you could either fix the generator to be smarter about copying over annotations or fix the definition of @c's type to allow it on fields. And I assume that Freezed doesn't copy across comments, so you can't just add an ignore comment in the original code that would then prevent the diagnostic from being reported in the generated code.

Correct. I could try and fix the issues one by one when users raise issues on the repository, but this is too tedious.

**Comment by rrousselGit** [comment]
In particular the most common complaints so far is `json_serializable`.

An example would be:

```dart
@freezed
class Example with _$Example {
  @JsonSerializable()
  factory Example(@JsonKey int value) = _Example;
}
```

**Comment by srawlins** [comment]
I agree with @bwilkerson. I think this is too broad of an ask. It would be more intentional, expressive, and precise to instead add  `// ignore` or `// ignore_for_file` comments in the freezed-generated code.

If `@JsonKey` is not allowed on a field, then it's not allowed on a field. I think the more proper solution would be to not annotate the field with an annotation that is not allowed on that field, or to allow that annotation on fields.

Additionally, I _think_ that freezed should have the information it would need to print a warning to users, "I am about to copy this annotation from a parameter to a field, but it is not supposed to be on a field. File a bug with the owner of the annotation."

**Comment by rrousselGit** [comment]
But as I said, the problem is not generated code. It is user defined code.

Generated code already is ignored from analysis. But there is nothing that can be done about user defined code.

**Comment by bwilkerson** [comment]
So why can't you add an ignore comment in the user defined code?

**Comment by creativecreatorormaybenot** [comment]
@bwilkerson you can, but it is a _horrible_ experience.

You will have to say that the way to use the package without warnings is adding an `// ignore: invalid_annotation_target` in every other file and line or disabling the lint altogether.

**Comment by bwilkerson** [comment]
I agree that that's not ideal.

But the more I think about this the more convinced I am that the problem isn't with the way `@Target` is defined or with the way it's being reported, but with the way it's being used. Either the annotation should be marked as being allowed in places where it currently isn't marked as being allowed, or it shouldn't be used in those places.

What am I missing?
